### PR TITLE
AST scopes/binding to names mapping.

### PR DESCRIPTION
**Pull request by yurydelendik**
Associated Issue: https://github.com/devtools-html/debugger.html/issues/3427

### Summary of Changes

* merges parsed scopes in generated source with names information from source maps

### Test Plan

See https://github.com/devtools-html/debugger.html/pull/3817

### CODE REVIEW

File: packages/devtools-source-map/src/source-map.js [hunk 1]
```diff
@@ -182,6 +184,108 @@ async function getOriginalLocation(location: Location) : Promise<Location> {
...
+    return null;
+  }
+
+  return generatedSourceScopes.map(scope => {
+    // The scope has bindings that contains positions of reference to the
+    // generated name (that may has different original name). Going over all
+    // these positions, check what original name is listed there. To make things
+    // more interesting the same generated name can potentially be used for
+    // different original names.
+    let trackedBindings: { [string]:
```
Comment (jasonLaster) [inline packages/devtools-source-map/src/source-map.js]: This type should be defined above:

`let trackedBindings: TrackedBindings`

File: packages/devtools-source-map/src/source-map.js [hunk 2]
```diff
@@ -182,6 +184,108 @@ async function getOriginalLocation(location: Location) : Promise<Location> {
...
+      {
+        name: string,
+        bestLocation: Location
+      }
+    } = Object.create(null);
+    Object.keys(scope.bindings).forEach(name => {
+      const positions = scope.bindings[name];
+      positions.forEach(position => {
```
[the remark this answers is not in the document]
Comment (jasonLaster) [reply packages/devtools-source-map/src/source-map.js]: `positions.reduce` might need the initial value `positions.reduce(position => findBestLocations(position), {}`

File: packages/devtools-source-map/src/source-map.js (hunk 1 above)
Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: done

[the remark this answers is not in the document]
Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: done

File: packages/devtools-source-map/src/source-map.js [hunk 3]
```diff
@@ -182,6 +184,108 @@ async function getOriginalLocation(location: Location) : Promise<Location> {
...
+          ) {
+            trackedBinding.name = name;
+            trackedBinding.bestLocation = position;
+          }
+        }
+      });
+    });
+
+    let bindings: { [string]: string } = Object.create(null);
+    Object.keys(trackedBindings).forEach(name => {
```
[the remark this answers is not in the document]
Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: fixed here and above.

File: packages/devtools-source-map/src/source-map.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: find best location shall be performed across all generated names -- I reviewed summary comment before algorithm and provided example why things are done the way they are.

File: packages/devtools-source-map/src/source-map.js [hunk 4]
```diff
@@ -182,6 +189,140 @@ async function getOriginalLocation(location: Location) : Promise<Location> {
...
+    !originalPositionTest ||
+    originalPositionTest.column !== originalPosition.column ||
+    originalPositionTest.line !== originalPosition.line
+  ) {
+    return null;
+  }
+  return originalPosition.name;
+}
+
+async function getLocationScopes(
```
Comment (jasonLaster) [inline packages/devtools-source-map/src/source-map.js]: I think `getLocationScope` is sufficiently complex to justify its own module with some helper methods.

```js

function appendTrackedBinding() {}

function appendBinding() {}

function getScopeBindings() {}

function getLocationScope() {

  generatedSourceScopes.map(scope => {
    // ...
    const trackedBindings = bindings.reduce(appendTrackedBinding, {})
    const bindings = trackedBindings.reduce(appendBinding, {});
    return {
      type: scope.type,
      bindings,
    };
  })
}

module.exports = { getLocationScope}
```

Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: appendTrackedBinding and appendBinding extraction is not possible due to dependency on location/currentPosition.

Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: getLocationScope also depends on _getSourceMap

Comment (jasonLaster) [reply packages/devtools-source-map/src/source-map.js]: we might need to do a bit more splitting to share `_getSourceMap` between modules.

we also might need to pass in the closure state:

```js
bindings.reduce(binding => appendTrackedBinding(binding, location, ...), {})
```

Comment (yurydelendik) [reply packages/devtools-source-map/src/source-map.js]: Did not extracted _getSourceMap -- just exposed it from ./source-map

Moved scopes into separate file and refactored to have appendTrackedBinding.

**Comment by jasonLaster** [comment]
I think it would be nice to switch to using a fixtures:

1. we can visualize the mappings 
2. easily regenerate them
3. be sure we didn't hardcode a bug in the mapping

Here is a [commit] where i setup uglify to make to do the minifying.

[commit]: https://github.com/devtools-html/devtools-core/commit/0e6b1da757999ff9054b40e0199f1dba794141f0

File: packages/devtools-source-map/src/scopes.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        generatedName,
+        nameLocation,
+        map,
+        currentPosition
+      ),
+    trackedBindings
+  );
+}
+
+async function getLocationScopes(
```
Comment (tromey) [inline packages/devtools-source-map/src/scopes.js]: I'd like jsdoc for all new exported functions.  Lack of documentation is one of the bigger problems in this entire module already.

File: packages/devtools-map-bindings/src/tests/scopes.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      line: 20,
+      column: 5
+    },
+    source: "foo.js",
+    original: {
+      line: 50,
+      column: 1
+    },
+    name: "zero"
+  });
```
Comment (jasonLaster) [inline packages/devtools-map-bindings/src/tests/scopes.js]: we can clean this up after we add the parser code and can improve the test in one pass.

**Pull request merged**